Rework dialer and listener to not assume anything at startup

klochowicz opened the pull request:
In short, supervised actors can be restarted at any time. If they subscribe to changes in the Endpoint, they first need to figure out the current state at startup. 
Establishing this assumption allowed to rework and simplify the code of these actors. 

See commit messages for details.

klochowicz committed:
Poll for connected peers when dialer connects

If a dialer is restarted, it might lose the information whether the endpoint is
connected; in other words, we cannot rely on the subscribed information
unless we build current information at startup.

Simplify the implementation by removing `tasks` and `connected` state; we now
try to connect, and if it doesn't work, then we shut down the actor.
Note that it is not necessary to subscribe for `ConnectionEstablished` message
anymore, as we poll for that information once to verify that request has worked.

klochowicz committed:
Rework listener to not need temporary state

Rework listener in a similar fashion to dialer - stop the actor only if the
listening request was not processed in time.

Note: We don't need to subscribe to ListenerAdded message; listener polls for
that manually at startup.

klochowicz committed:
Adjust stale comments in Listener and Dialer

The mechanism of how these actors work has changed significantly; comments
should reflect that.

da-kami commented on xtra-libp2p/src/dialer.rs:
```diff
@@ -96,18 +84,42 @@ impl Actor {
+        // Only check the connection again after it had enough time to be established
+        tokio::time::sleep(CONNECTION_TIMEOUT).await;
+
+        anyhow::ensure!(
```
Should this really be an ensure? This changes the constraints. I would have seen it as a `warn` log if we could not establish the connection and I would keep a faster retry from the dialer (i.e. 5 secs)

klochowicz replied:
what's the changed constraint here? 

- the original behaviour: we wait, then we check, if we're not connected we stop the actor
- subscriber, v1: we spawn a task in the background that checks after the same timeout, and if we're not connected by then, we stop the actor.
- this PR: we wait, check, and if we're not connected we stop the actor (same thing really).

What changed a bit is how the actor is stopped, and that might have caused the confusion:
- original version: we dispatch an internal message to stop the actor
- subscribers, v1: we have a bunch of `if let Err(e)` that calls the `stop_with_error()` method
- this PR: I extracted an internal method, so I could use `if let Err(e)` once for the resulting anyhow error.

IMHO the second ensure is the crucial one, because this is the one that covers trying again. I don't see need for faster retry? (if you want, we can drop the alignment of the timeouts, so it always just tries every 5s, I can be convinced to that one :) )

da-kami commented on xtra-libp2p/src/dialer.rs:
```diff
@@ -2,38 +2,37 @@ use crate::endpoint;
 
 /// If we're not connected by this time, stop the actor.
-pub const CONNECTION_TIMEOUT: Duration = Duration::from_secs(5);
+pub const CONNECTION_TIMEOUT: Duration = Duration::from_secs(20);
```
I don't think this is necessary. We should potentially try more often then every `20` seconds. 

If we trigger this and the endpoint is still waiting the endpoint will just return an error hat we are still trying to connect.
But it's fine for the dialer to check again after 5 secs.

Aligning this could mean that retrying to connect takes 20 seconds.

Consider this:

1. The dialer triggers a `Connect` in the endpoint.
2. The endpoint fails to connect after 6 seconds. 
3. The dialer restarts only after 20... 

If the dialer would restart with the 5 secs it would catch such a scenario faster, and it should not do harm.

klochowicz replied:
I will revert the last commit, as 5s was catching the scenario faster; as such, it probably should have not been included in this PR.

we can add the more elaborate reconnection scenario in a separate PR.

da-kami reviewed (commented):
My comments are not blocking, but I do think we should not align the timeouts between dialer and endpoint. This was different for a reason.

The concept of just using `GetConnectionStats` and `ConnectionDropped`: ack
The concept of using `send_async_safe` instead of tasks: ack

da-kami commented on xtra-libp2p/src/dialer.rs:
```diff
@@ -96,18 +84,42 @@ impl Actor {
-            )
+    async fn handle(&mut self, _msg: Dial, ctx: &mut xtra::Context<Self>) {
+        if let Err(e) = self.dial().await {
+            self.stop_with_error(Error::Failed { source: e }, ctx);
```
Maybe add a log with the error here so we know what failed :)
Or let the supervisor log it.

klochowicz replied:
added inside `stop_with_error` function, both on dialer and listener.

klochowicz committed:
Remove unused message handlers

We moved away from dispatching this as an internal messages

klochowicz committed:
Add log when stopping an actor with error

The actor will be restarted by a supervisor, but we might like to see more
information in the logs if debugging.
